docs/concepts/components/children.md

```diff
@@ -0,0 +1,230 @@
+what type of children the component has. In such cases, the below example will
+suffice.
+
+```rust
+use yew::prelude::*;
+
+#[derive(Properties, Clone)]
+pub struct ListProps {
+    #[prop_or_default]
+    pub children: Children,
+}
+
+pub struct List {
+    props: ListProps,
+}
+
+impl Component for List {
+    type Properties = ListProps;
+    // ...
+
+    fn view(&self) -> Html {
+        html! {
+            <div class="list">
+                { for self.props.children.iter() }
+            </div>
+        }
+    }
+}
+```
+
+## Advanced usage
+
+There are some cases where you wouldn't want to use the `Children` type
+and instead use an alternative method:
+
+ - You want to only allow components of a specific type to be used as children for your component.
+ - You want to share state between a component and its children.
+
+### Typed children
+
+In cases where you want one type of component to be passed as children to your component,
+you can use `yew::html::ChildrenWithProps<T>`.
+
+```rust
+use yew::html::ChildrenWithProps;
+use yew::prelude::*;
+
+// ...
+
+#[derive(Properties, Clone)]
+pub struct ListProps {
+    #[prop_or_default]
+    pub children: ChildrenWithProps<Item>,
+}
+
+pub struct List {
+    props: ListProps,
+}
+
+impl Component for ListProps {
+    type Properties = ListProps;
+    // ...
+
+    fn view(&self) -> Html {
+        html! {
+            <div class="list">
+                { for self.props.children.iter() }
+            </div>
+        }
+    }
+}
+```
+
+Of course, sometimes you might need to restrict the children to a few different
+components. In these cases, you have to get a little more hands-on with Yew.
+
+The [`derive_more`](https://github.com/JelteF/derive_more) crate is used here
+for better ergonomics. If you don't want to use it, you can manually implement
+`From` for each variant.
+
+```rust
+use yew::prelude::*;
+use yew::html::ChildrenRenderer;
+use yew::virtual_dom::{ VChild, VComp };
+
+// `derive_more::From` implements `From<VChild<Primary>>` and
+// `From<VChild<Secondary>>` for `Item` automatically!
+#[derive(Clone, derive_more::From)]
+pub enum Item {
+    Primary(VChild<Primary>),
+    Secondary(VChild<Secondary>),
+}
+
+// Now, we implment `Into<Html>` so that yew knows how to render `Item`.
+impl Into<Html> for Item {
+    fn into(self) -> Html {
+        match self {
+            Self::Primary(child) => child.into(),
+            Self::Secondary(child) => child.into(),
+        }
+    }
+}
+
+#[derive(Properties, Clone)]
+pub struct ListProps {
+    #[prop_or_default]
+    pub children: ChildrenRenderer<Item>,
+}
+
+pub struct List {
+    props: ListProps,
+}
+
+impl Component for List {
+    type Properties = ListProps;
+    // ...
+
+    fn view(&self) -> Html {
+        html! {
+            <div class="list">
+                { for self.props.children.iter() }
+            </div>
+        }
+    }
+}
+```
+
+### Sharing properties
```

Comment:
Instead of keeping you waiting even longer let me just outline what I consider to be the problem with this section and particularly the example below.

So first of all, the example kind of makes it sound like this is what you need to do to access the properties of the component when in fact it's already possible with the example from above (You can always access the properties through `VChild<T>.props` which is what's contained in `ChildrenWithProps<T>`).

A more fitting description of this section is that it's about how to limit the children to a certain type of `Component::Properties` rather than just the `Component` type from above.
In the example below any component which either has `type Properties = PrimaryProps` or `type Properties = SecondaryProps` is accepted but that brings me to the problem with it. It accepts **any** component type with either of the property types **but** it completely erases the component type itself. It always turns into a `<Primary />` or `<Secondary />` component so if we try to create our own little component which accepts these properties we would end up very confused.
I think I don't need to go into why this isn't great.

To avoid this issue `Item` would have to keep track of the original `VChild<T>` (probably in the form of `VComp` though because it isn't generic) but then we need to clone the props instead of moving them out of the `VChild` and at that point we can no longer manipulate them.
A better alternative might be to create a trait, implement it on the property types we wish to accept, and then keep a trait object `Box<dyn TheTrait>` in `Item` but that's as confusing as it probably sounds. Generally, I think there's no general solution for this. It all depends on the use case.

Unless you can think of a concrete example (or maybe a way to implement this properly) I think it's best to just remove this section for now. We can always pick it back up if it becomes a requested feature in the future.

TL;DR: let's speed things up by getting rid of this entire section, then we can merge this and move on :)

Reply:
Sounds good to me :+1: